src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+                    localChangedProject = e.ProjectId;
+                    break;
+
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionChanged:
+                case WorkspaceChangeKind.SolutionCleared:
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                    localSolutionChanged = true;
+                    break;
+            }
+
+            lock (_gate)
+            {
+                // Cancel any existing update task.
+                _tokenSource.Cancel();
+                _tokenSource = new CancellationTokenSource();
+
+                _solutionChanged |= localSolutionChanged;
+                if (localChangedProject != null)
+                {
+                    _changedProjects.Add(localChangedProject);
+                }
+
+                // Because we may get a lot of events, we attempt to throttle things so that
+                // we only update our nuget information once a second has passed.
+                var cancellationToken = _tokenSource.Token;
+                Task.Delay(TimeSpan.FromSeconds(1), cancellationToken)
+                    .ContinueWith(_ => OnWorkspaceChangedOnForeground(cancellationToken), cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, this.ForegroundTaskScheduler);
+            }
+        }
+
+        private void OnWorkspaceChangedOnForeground(CancellationToken cancellationToken)
+        {
+            this.AssertIsForeground();
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+
+            if (_workspace == null || _packageInstallerServices == null)
+            {
+                return;
+            }
+
+            var solution = _workspace.CurrentSolution;
+            ProjectId projectId;
+            bool clearAllData;
+            DequeueNextProject(solution, out projectId, out clearAllData);
+
+            if (clearAllData)
+            {
+                // Just clear out everything we have and start over in the case of a solution change.
+                _projectToInstalledPackageAndVersion.Clear();
+            }
+
+            if (projectId == null)
+            {
+                // No project to process, nothing to do.
+                return;
+            }
+
+            // Process this single project.
+            ProcessProjectChange(solution, projectId);
+
+            // After processing this single project, yield so the foreground thread
+            // can do more work.  Then go and loop again so we can process the 
+            // rest of the projects.
+            Task.Factory.SafeStartNew(
+                () => OnWorkspaceChangedOnForeground(cancellationToken), cancellationToken, ForegroundTaskScheduler);
+        }
+
+        private void DequeueNextProject(Solution solution, out ProjectId projectId, out bool clearAllData)
```

Comment:
`clearAllData` feels like a premature optimization to me with little benefit. It's used to preemptively clear out `_projectToInstalledPackageAndVersion`, but you have to go visit each project to repopulate them independently anyway so it could just be done then. If you're worried about potentially "stale" info being kept around for those moments where you yield the UI thread, then you'll still have a problem for the case of 9 projects changing in a 10 project solution -- we won't preemptively clear `_projectToInstalledPackageAndVersion` and the stale info will be kept until it's replaced for each project, one at a time.

And then this could just return the ProjectId that you dequeued directly. :smile: